components/raftstore/src/store/fsm/apply.rs

```diff
@@ -2937,13 +2955,53 @@ where
         ))
     }
 
+    fn try_compact_log(
+        &mut self,
+        voter_replicated_index: u64,
+    ) -> Result<Option<TaskRes<EK::Snapshot>>> {
+        PEER_ADMIN_CMD_COUNTER.compact.all.inc();
+        let first_index = entry_storage::first_index(&self.apply_state);
+
+        if self.is_merging {
+            info!(
+                "in merging mode, skip compact";
+                "region_id" => self.region_id(),
+                "peer_id" => self.id(),
+                "voter_replicated_index" => voter_replicated_index,
+            );
+            return Ok(None);
+        }
+
+        match self.pending_cmds.pop_compact(voter_replicated_index) {
+            Some(cmd) => {
+                // compact failure is safe to be omitted, no need to assert.
+                compact_raft_log(&self.tag, &mut self.apply_state, cmd.index, cmd.term)?;
+                PEER_ADMIN_CMD_COUNTER.compact.success.inc();
+                Ok(Some(TaskRes::Compact {
+                    state: self.apply_state.get_truncated_state().clone(),
```

Comment:
When will `apply_state` be persisted?

Reply:
missed :(                 I plan to persist before notifying the store thread